### PR TITLE
Add ability to use OCI hooks from config for runc workers

### PR DESCRIPTION
**Pull request by dchw**
Relies on https://github.com/earthly/earthly/pull/2289 to be present in the image.

### CODE REVIEW

**Comment by alexcb** [comment]
can you update the buildkit sha in earthly to reflect this change, in order to show all our tests pass.

**Review by vladaionescu — APPROVED** [review]

**Review by alexcb — COMMENTED** [review]
```
        	            	failed to create shim task: OCI runtime create failed: runc create failed: error running hook #0: fork/exec /bin/oom-adjust.sh: no such file or directory: unknown
```

looks like this broke a lot of tests.

**Comment by dchw** [comment]
> can you update the buildkit sha in earthly to reflect this change, in order to show all our tests pass.

Do I then have to back out the change before merge, so I can update it to the one in main later?

**Comment by alexcb** [comment]
> > can you update the buildkit sha in earthly to reflect this change, in order to show all our tests pass.
> 
> Do I then have to back out the change before merge, so I can update it to the one in main later?

No need to back out any changes, just going off what we have documented under https://github.com/earthly/earthly/blob/main/CONTRIBUTING.md#updates-to-buildkit-or-fsutil
we need to show that these changes are compatible with earthly; so in your specific case, you should run

```
earthly +update-buildkit --BUILDKIT_GIT_ORG=<git-user-or-org> --BUILDKIT_GIT_SHA=08d76353e768d6c648f22d6a3e198925d2ab6864
```

under your earthly branch in https://github.com/earthly/earthly/pull/2289 to show that these all work together.

Then once the buildkit PR is merged, you'll need to change your earthly PR to reflect the sha post-merge.

**Comment by dchw** [comment]
> looks like this broke a lot of tests.

Yep, this was why I wanted to move the hook script in here. But now theres a better way.

**Comment by dchw** [comment]
This includes a rebase off `earthly-main`. There is still some additional work to bring our fork up to date? This will wait to merge until thats in and I can rebase again.

File: cmd/buildkitd/config/config.go [hunk 1]
```diff
@@ -94,6 +94,9 @@ type OCIConfig struct {
 
 	// MaxParallelism is the maximum number of parallel build steps that can be run at the same time.
 	MaxParallelism int `toml:"max-parallelism"`
+
+	// Hooks are things you can run during any phase of the OCI container runtime lifecycle.
```
Comment (alexcb) [inline cmd/buildkitd/config/config.go]:
```suggestion
	// earthly-specific: Hooks are things you can run during any phase of the OCI container runtime lifecycle.
```

File: worker/runc/runc.go [hunk 2]
```diff
@@ -67,6 +67,7 @@ func NewWorkerOpt(root string, snFactory SnapshotterFactory, rootless bool, proc
 		ApparmorProfile:     apparmorProfile,
 		TracingSocket:       traceSocket,
 		DefaultCgroupParent: defaultCgroupParent,
+		Hooks:               hooks,
```
Comment (alexcb) [inline worker/runc/runc.go]: the `NewWorkerOpt` function might change overtime (in upstream commits), which will increase the odds of a conflict. It'll make maintaining this easier if we add an earthly-specific comment:

```suggestion
		Hooks:               hooks, //earthly-specific
```

**Review by alexcb — APPROVED** [review]
looks good assuming all the earthly checks pass (no need to change the sha to reflect the two comment changes).

**Pull request merged**